src/accel/RunAction.cc

```diff
@@ -59,12 +83,122 @@ void RunAction::EndOfRunAction(const G4Run*) {}
+            ->GetNavigatorForTracking()
+            ->GetWorldVolume());
+
+    auto imported = load_geant_data();
+    CELER_LOG_LOCAL(info) << "loaded data: "
+                          << (imported ? "success" : "failure");
+
+    CoreParams::Input params;
+
+    // Create action manager
+    {
+        params.action_reg = std::make_shared<ActionRegistry>();
+    }
+
+    // Load geometry
+    {
+        params.geometry = std::make_shared<GeoParams>(options_->geometry_file);
+        if (!params.geometry->supports_safety())
+        {
+            CELER_LOG(warning)
+                << "Geometry contains surfaces that are "
+                   "incompatible with the current ORANGE simple "
+                   "safety algorithm: multiple scattering may "
+                   "result in arbitrarily small steps";
+        }
+    }
+
+    // Load materials
+    {
+        params.material = MaterialParams::from_import(imported);
+    }
+
+    // Create geometry/material coupling
+    {
+        params.geomaterial = GeoMaterialParams::from_import(
+            imported, params.geometry, params.material);
+    }
+
+    // Construct particle params
+    {
+        params.particle = ParticleParams::from_import(imported);
+    }
+
+    // Construct cutoffs
+    {
+        params.cutoff = CutoffParams::from_import(
+            imported, params.particle, params.material);
+    }
+
+    // Load physics: create individual processes with make_shared
+    {
+        PhysicsParams::Input input;
+        input.particles       = params.particle;
+        input.materials       = params.material;
+        input.action_registry = params.action_reg.get();
+
+        input.options.linear_loss_limit = imported.em_params.linear_loss_limit;
+        input.options.secondary_stack_factor = options_->secondary_stack_factor;
+
+        {
+            ProcessBuilder::Options opts;
+            ProcessBuilder          build_process(
+                imported, opts, params.particle, params.material);
+
+            std::set<ImportProcessClass> all_process_classes;
+            for (const auto& p : imported.processes)
+            {
+                all_process_classes.insert(p.process_class);
+            }
+            for (auto p : all_process_classes)
+            {
+                input.processes.push_back(build_process(p));
+            }
+        }
+
+        params.physics = std::make_shared<PhysicsParams>(std::move(input));
+    }
+
+    // TODO: different along-step kernels
+    {
+        // Create along-step action
+        auto along_step = AlongStepGeneralLinearAction::from_params(
+            params.action_reg->next_id(),
+            *params.material,
+            *params.particle,
+            *params.physics,
+            imported.em_params.energy_loss_fluct);
+        params.action_reg->insert(along_step);
+    }
+
+    // Construct RNG params
+    {
+        params.rng
+            = std::make_shared<RngParams>(CLHEP::HepRandom::getTheSeed());
```

Comment:
Ooh I didn't think about this: `getTheSeed` is thread-local (and we want our seeds to be stream-local!) but the `RngParams` is shared... hm...

Reply:
Actually even `CLHEP::HepRandom::setTheSeed` looks to be thread-local... so perhaps @whokion could help shed some light on what's expected to be thread/task-local...